### ISSUE TITLE
<string.h>: memmove intrinsic bad codegen

### ISSUE THREAD

**Issue by blat-blatnik** [microsoft/STL#2610]
**Describe the bug**

MSVC generates faulty code for small `memmove` in release mode - the test case below works fine in debug mode MSVC, clang -O3, gcc -O3, but fails in MSVC release mode.

**Command-line test case**
```c
C:\Temp>type repro.c
#undef NDEBUG
#include <assert.h>
#include <string.h>
#include <stdio.h>
int main(void) {
	char a[5] = { 1, 2, 3, 4, 0 };
	printf("%d %d %d %d %d\n", a[0], a[1], a[2], a[3], a[4]); // 1 2 3 4 0
	memmove(a + 1, a, 4);
	printf("%d %d %d %d %d\n", a[0], a[1], a[2], a[3], a[4]); // 1 1 1 2 3 !!!
	assert(a[0] == 1 && a[1] == 1 && a[2] == 2 && a[3] == 3 && a[4] == 4);
}

C:\Temp>cl /O2 .\repro.c
Microsoft (R) C/C++ Optimizing Compiler Version 19.31.31104 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.c
Microsoft (R) Incremental Linker Version 14.31.31104.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:repro.exe
repro.obj

C:\Temp>.\repro.exe
1 2 3 4 0
1 1 1 2 3
Assertion failed: a[0] == 1 && a[1] == 1 && a[2] == 2 && a[3] == 3 && a[4] == 4, file repro.c, line 10
```

**Expected behavior**
The test case prints `1 2 3 4 0\n1 1 2 3 4\n` and then terminates without triggering the assertion.

**STL version**
```
Microsoft Visual Studio Community 2022 (64-bit) - Current
Version 17.1.1
```

**Additional context**
This is probably an optimization bug. If I call memmove through a function pointer, or make the array volatile, the test passes.

I realize that this is probably a compiler bug and not an STL bug, but I can't log into my microsoft account to file a bug on https://developercommunity.visualstudio.com. It seemed like a nasty enough bug that I had to report it somewhere - I hope someone here can bring it up over there.

From looking at the codegen on compiler explorer, it looks like this bug was introduced in version 19.31 of the compiler. In version 19.30 and earlier the codegen looks fine.

**Comment by fsb4000** [comment]
I think they have fixed it in 17.2 preview 2. (current preview)

I can reproduce your example with 17.1:
![изображение](https://user-images.githubusercontent.com/4289847/158852922-f1a17857-a285-43b9-9f4c-ff19f8fd1b8e.png)

And 17.2 p2:
![изображение](https://user-images.githubusercontent.com/4289847/158853018-0dd5ebbf-e2a3-4fd3-8856-94703b431366.png)

**Comment by StephanTLavavej** [comment]
Yep, this was a compiler bug that was fixed by internal MSVC-PR-365216 "Fix symbol CSE with overlapping source/dest stores". The fix was backported to VS 2022 17.0.2 as the 17.0 series is a long-term support release, see https://docs.microsoft.com/en-us/visualstudio/releases/2022/release-notes-v17.0#issues-addressed-in-this-release-of-1702 :

> Issues Addressed in this release of 17.0.2
> * Fix for silent-bad codegen with memmove-like operations with overlapping data
